addon/services/toast.ts

```diff
@@ -0,0 +1,138 @@
+import { later } from '@ember/runloop';
+import Service from '@ember/service';
+import { isEmpty } from '@ember/utils';
+
+const TIMEOUT = 5000;
+
+enum ToastType {
+  INFO = 'info',
+  WARNING = 'warning',
+  ERROR = 'error',
+  SUCCESS = 'success'
+}
+
+type ToastOptions = {
+  closeButton?: true;
+  tapToDismiss?: boolean;
+  onclick?: Function | null;
+};
+
+const DEFAULT_OPTIONS: ToastOptions = Object.freeze({
+  closeButton: true,
+  onclick: null,
+  tapToDismiss: true
+});
+
+function delegate(ctx: Toast, type: ToastType): Function {
+  return (message: string, title: string, opts: ToastOptions = DEFAULT_OPTIONS): void => {
+    if (isEmpty(message) && isEmpty(title)) return;
+
+    console.log(opts);
+    ctx.buildToast(type, message, title, { ...DEFAULT_OPTIONS, ...opts });
+  };
+}
+
+export default class Toast extends Service {
+  private _toasts: Element[] = [];
+
+  success = delegate(this, ToastType.SUCCESS);
+  error = delegate(this, ToastType.ERROR);
+  warning = delegate(this, ToastType.WARNING);
+  info = delegate(this, ToastType.INFO);
+
+  buildToast(type: ToastType, message: string, title: string, opts: ToastOptions): void {
+    const container = this._buildContainer();
+    console.log(opts);
+
+    const toast = document.createElement('div');
+    toast.classList.add('toast', 'toast--hidden', `toast-${type}`);
+
+    const titleContainer = document.createElement('div');
+    titleContainer.classList.add('toast-title');
+    titleContainer.textContent = title;
+
+    const messageContainer = document.createElement('div');
+    messageContainer.classList.add('toast-message');
+    messageContainer.textContent = message;
+
+    let toastChildren: Element[] = [titleContainer, messageContainer];
+
+    if (opts.closeButton) {
+      const closeButton = document.createElement('button');
+
+      closeButton.classList.add('toast-close-button');
+      closeButton.innerHTML = '<i class="fas fa-times text-size-5"></i>';
+      closeButton.addEventListener('click', this._onToastClose.bind(this), false);
+
+      toastChildren.push(closeButton);
+    }
+
+    if (opts.tapToDismiss) {
+      toast.addEventListener('click', () => this._destroyToast(toast), false);
+    }
+
+    toastChildren.forEach((x) => toast.append(x));
+
+    if (container.querySelector('.toast')) {
+      container.insertBefore(toast, container.querySelector('.toast'));
+    } else {
+      container.append(toast);
+    }
+
+    later(
+      this,
+      () => {
+        toast.classList.remove('toast--hidden');
+        toast.classList.add('toast--visible');
+      },
+      500
+    );
```

Comment:
just to understand, why this ?

Reply:
because we can't animate an element being added to the dom (nor can we add `transition` on the CSS `display` rule). So I add it to the dom first, then animate the visibility.

Reply:
Nice, didn't see the animation :+1: